kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,958 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2024.
+
+//! Dynamic Process Loader for OTA application loads and updates
+//!
+//! These functions facilitate dynamic application loading and process
+//! creation during runtime without requiring the user to restart the
+//! device.
+
+use core::cell::Cell;
+use core::cmp;
+
+use crate::capabilities::ProcessManagementCapability;
+use crate::config;
+use crate::create_capability;
+use crate::debug;
+use crate::hil::nonvolatile_storage::{NonvolatileStorage, NonvolatileStorageClient};
+use crate::kernel::Kernel;
+use crate::platform::chip::Chip;
+use crate::process::ProcessId;
+use crate::process::{self, Process, ShortId};
+use crate::process_binary::{ProcessBinary, ProcessBinaryError};
+use crate::process_loading::ProcessLoadError;
+use crate::process_policies::ProcessFaultPolicy;
+use crate::process_standard::ProcessStandard;
+use crate::utilities::cells::{MapCell, OptionalCell, TakeCell};
+use crate::ErrorCode;
+
+const TBF_HEADER_LENGTH: usize = 16;
+pub const BUF_LEN: usize = 512;
+const MAX_PROCS: usize = 10; //need this to store the start addresses of processes to write padding
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum NonvolatileCommand {
+    UserspaceRead,
+    UserspaceWrite,
+}
+
+#[derive(Clone, Copy)]
+pub enum NonvolatileUser {
+    Client,
+    Kernel,
+}
```

Comment:
I think it would be better to make this a traditional `State` enum. It's a little confusing what client and kernel mean here. Instead I think we just need to track the state (e.g. writing app, writing padding, etc).

Reply:
Added State enum.